File: crates/library/src/entity/album.rs
use serde::{Deserialize, Serialize};

#[derive(sqlx::FromRow, Default, Serialize, Deserialize, Clone)]
pub struct Album {
    pub id: String,
    pub title: String,
    pub artist: String,
    pub year: u32,
    pub year_string: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub album_art: Option<String>,
    pub md5: String,
    pub artist_id: String,
}

impl Into<rockbox_search::rockbox::search::v1alpha1::Album> for Album {
    fn into(self) -> rockbox_search::rockbox::search::v1alpha1::Album {
        rockbox_search::rockbox::search::v1alpha1::Album {
            id: self.id,
            title: self.title,
            artist: self.artist,
            year: self.year,
            year_string: self.year_string,
            album_art: self.album_art,
            md5: self.md5,
            artist_id: self.artist_id,
        }
    }
}

impl Into<rockbox_search::rockbox::search::v1alpha1::LikedAlbum> for Album {
    fn into(self) -> rockbox_search::rockbox::search::v1alpha1::LikedAlbum {
        rockbox_search::rockbox::search::v1alpha1::LikedAlbum {
            id: self.id,
            title: self.title,
            artist: self.artist,
            year: self.year,
            year_string: self.year_string,
            album_art: self.album_art,
            md5: self.md5,
            artist_id: self.artist_id,
        }
    }
}

File: sdk/elixir/lib/rockbox/album.ex
defmodule Rockbox.Album do
  @moduledoc "An album in the music library. `:tracks` may be partially populated."

  @type t :: %__MODULE__{
          id: String.t(),
          title: String.t(),
          artist: String.t(),
          year: integer(),
          year_string: String.t(),
          album_art: String.t() | nil,
          md5: String.t(),
          artist_id: String.t(),
          copyright_message: String.t() | nil,
          tracks: [Rockbox.Track.t()]
        }

  defstruct [
    :id,
    :title,
    :artist,
    :year,
    :year_string,
    :album_art,
    :md5,
    :artist_id,
    :copyright_message,
    tracks: []
  ]
end
